In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
PATH = "/content/drive/MyDrive/PRI_Visualizacion/"

In [ ]:
#https://www.kaggle.com/datasets/gpreda/covid-world-vaccination-progress
# Read the CSV file
Vacunas_por_pais_df = pd.read_csv(PATH+"country_vaccinations.csv")

In [ ]:
# Show the first rows of the DataFrame
print("Primeras filas del DataFrame:")
print(Vacunas_por_pais_df.head())

# Get a summary of the descriptive statistics
print("\nResumen de las estadísticas descriptivas:")
print(Vacunas_por_pais_df.describe())

# Get general information about the DataFrame
print("\nInformación del DataFrame:")
print(Vacunas_por_pais_df.info())

Primeras filas del DataFrame:
       country iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per

In [ ]:
# Drop the "daily_vaccinations_raw" and "source_website" columns
Vacunas_por_pais_df.drop(columns=["daily_vaccinations_raw", "source_website"], inplace=True)

In [ ]:
# Print the column names of each DataFrame
print("Column names:")
print(Vacunas_por_pais_df.columns)

Column names:
Index(['country', 'iso_code', 'date', 'total_vaccinations',
       'people_vaccinated', 'people_fully_vaccinated', 'daily_vaccinations',
       'total_vaccinations_per_hundred', 'people_vaccinated_per_hundred',
       'people_fully_vaccinated_per_hundred', 'daily_vaccinations_per_million',
       'vaccines', 'source_name'],
      dtype='object')


In [ ]:
# Show the number of rows of each DataFrame
num_filas = Vacunas_por_pais_df.shape[0]
print(f"Número de filas en el DataFrame: {num_filas}")

Número de filas en el DataFrame: 86512


In [ ]:
# Count null values per column
valores_nulos_por_columna = Vacunas_por_pais_df.isnull().sum()

# Count empty values per column
valores_vacios_por_columna = (Vacunas_por_pais_df == '').sum()

print("Valores nulos por columna:")
print(valores_nulos_por_columna)

print("\nValores vacíos por columna:")
print(valores_vacios_por_columna)

Valores nulos por columna:
country                                    0
iso_code                                   0
date                                       0
total_vaccinations                     42905
people_vaccinated                      45218
people_fully_vaccinated                47710
daily_vaccinations                       299
total_vaccinations_per_hundred         42905
people_vaccinated_per_hundred          45218
people_fully_vaccinated_per_hundred    47710
daily_vaccinations_per_million           299
vaccines                                   0
source_name                                0
dtype: int64

Valores vacíos por columna:
country                                0
iso_code                               0
date                                   0
total_vaccinations                     0
people_vaccinated                      0
people_fully_vaccinated                0
daily_vaccinations                     0
total_vaccinations_per_hundred         0
people_vaccinated_

In [ ]:
# Drop rows with more than 50% null values
#Vacunas_por_pais_df = Vacunas_por_pais_df.dropna(thresh=len(Vacunas_por_pais_df.columns) // 2)

In [ ]:
# Convert the 'date' column to date format
Vacunas_por_pais_df['date'] = pd.to_datetime(Vacunas_por_pais_df['date'])

In [ ]:
print(Vacunas_por_pais_df.head())

       country iso_code       date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG 2021-02-22                 0.0                0.0   
1  Afghanistan      AFG 2021-02-23                 NaN                NaN   
2  Afghanistan      AFG 2021-02-24                 NaN                NaN   
3  Afghanistan      AFG 2021-02-25                 NaN                NaN   
4  Afghanistan      AFG 2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations  \
0                      NaN                 NaN   
1                      NaN              1367.0   
2                      NaN              1367.0   
3                      NaN              1367.0   
4                      NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             Na

In [ ]:
Vacunas_por_pais_df.to_csv(PATH + "country_vaccinations_procesado.csv", index=False)

**What do we know about its structure?**

The "COVID-19 World Vaccination Progress" dataset contains detailed data on COVID-19 vaccination at a global level, organized in a tabular format with 15 columns and 86,512 rows. Each row represents a daily record for a country, providing a broad set of data for temporal and geographic analysis. The structure includes both numerical and categorical variables.

**Who published it?**

This dataset was published on Kaggle by the user GABRIEL PREDA. Kaggle is a platform recognized for its extensive collection of datasets used by researchers and data scientists. Publication on Kaggle allows users to review and comment on the quality of the data, contributing to its reliability.

**What variables does it contain?**

The variables included in the dataset are:

**country:** The country where the vaccination data are recorded.

**iso_code:** ISO code of the country.

**date:** Date of the vaccination data.

**total_vaccinations:** Total number of doses administered.

**people_vaccinated:** Total number of people who have received at least one dose.

**people_fully_vaccinated:** Total number of people who have received all the required doses.

**daily_vaccinations_raw:** Daily number of doses administered (raw).

**daily_vaccinations:** Daily number of doses administered (7-day moving average).

**total_vaccinations_per_hundred:** Total number of doses administered per 100 people in the population.

**people_vaccinated_per_hundred:** Total number of people vaccinated (at least one dose) per 100 people in the population.

**people_fully_vaccinated_per_hundred:** Total number of people fully vaccinated per 100 people in the population.

**daily_vaccinations_per_million:** Daily number of doses administered per million people in the population.

**vaccines:** Name of the vaccines used in the country.

**source_name:** Name of the data source.

**source_website:** Website of the data source.

**What is its internal structure like?**
The internal structure of the dataset is organized so that each row represents the vaccination data of a country on a specific date. The columns contain the values associated with that date and country, allowing a detailed analysis of vaccination progress over time and across different countries. The variables are related to one another mainly through the country and the date, allowing temporal and geographic tracking of the vaccination process.

**How are the variables related to one another?**

The variables are related mainly through the geographic (country) and temporal (date) dimensions. For example, the number of doses administered daily can be analyzed in relation to the cumulative total and the doses per 100 people to assess vaccination progress in different countries and periods. Categorical variables such as "country" and "vaccines" make it possible to segment the data for more specific analyses.

**Additional exploratory data:**

**First rows of the DataFrame:** Shows records from Afghanistan, with data starting on February 22, 2021.

**Descriptive statistics:** They show that there is wide variability in vaccinations, with a significant range between the minimum and the maximum number of doses administered.

**DataFrame information:**


Total number of records: 86,512

Total number of columns: 15


Columns with significant null values:


total_vaccinations: 42,905 null values

people_vaccinated: 45,218 null values

people_fully_vaccinated: 47,710 null values

daily_vaccinations_raw: 50,150 null values

daily_vaccinations_per_million: 299 null values



**Justification for the choice of the dataset:**

We chose this dataset because it provides a complete view of the progress of COVID-19 vaccination at a global level, a topic of great current relevance. The diversity of the variables and the richness of the data make it possible to carry out in-depth, comparative analyses across different countries and periods. In addition, the format of the dataset is suitable for a variety of data visualization techniques, which facilitates the understanding and communication of the findings.

**Interest of the dataset**

The dataset is interesting because:

It comes from a public and reliable source (Kaggle) with data reviewed by the community.

It contains a high density of numerical and categorical data, allowing a wide range of analyses.

It makes it possible to investigate the impact of the vaccination campaign at a global level and to compare the effectiveness and the reach across different countries.

It can be combined with other datasets to carry out more complete studies, such as the relationship between vaccination progress and the evolution of COVID-19 cases.
For more details about the dataset and its variables, you can visit the Kaggle page: COVID-19 World Vaccination Progress​ (Kaggle ML & Data Sci)​​ (Kaggle)​.

In [ ]:
!pip install pycountry_convert

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 45.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 231.6/231.6 kB 25.9 MB/s eta 0:00:00


In [3]:
df = pd.read_csv(PATH+'country_vaccinations_procesado.csv')

Question 1:

Which vaccines are the most used by country and year? Dynamic world map where the total number for each type of vaccine can be seen over each country. A color range representing the number.

In [4]:
import plotly.express as px

# Make sure the date column is of datetime type
df['date'] = pd.to_datetime(df['date'])

# Extract the year from the date column
df['year'] = df['date'].dt.year

# Filter the data by year
df_2020 = df[df['year'] == 2020]
df_2021 = df[df['year'] == 2021]
df_2022 = df[df['year'] == 2022]

# Function to create the interactive map
def create_choropleth(data, year, width=2000, height=1500):
    df_grouped = data.groupby(['country', 'vaccines', 'date'], as_index=False)['daily_vaccinations_per_million'].sum()
    df_grouped['date'] = df_grouped['date'].dt.strftime('%Y-%m-%d')  # Convert the dates to strings for Plotly
    fig = px.choropleth(df_grouped, locations='country', locationmode='country names',
                        color='daily_vaccinations_per_million', hover_name='country', hover_data=["vaccines", "daily_vaccinations_per_million"], animation_frame='date',
                        color_continuous_scale=px.colors.sequential.Plasma, title=f'Vacunas más usadas por país en {year}',
                        labels={'daily_vaccinations_per_million': 'Vacunación diaria por millón de habitantes', 'date': 'Fecha'})
    fig.update_layout(width=width, height=height)  # Update the layout with the desired size
    return fig

# Create the maps for each year
fig_2020 = create_choropleth(df_2020, 2020)
fig_2021 = create_choropleth(df_2021, 2021)
fig_2022 = create_choropleth(df_2022, 2022)

# Save the maps as HTML files
fig_2020.write_html(PATH+"vaccines_choropleth_2020.html")
fig_2021.write_html(PATH+"vaccines_choropleth_2021.html")
fig_2022.write_html(PATH+"vaccines_choropleth_2022.html")
fig_2020.show()
fig_2021.show()
fig_2022.show()


Output hidden; open in https://colab.research.google.com to view.

Question 2:

How has the monthly vaccination rate varied across different continents over time? Line charts for each country, showing the monthly evolution of vaccinations, where the X axis represents the interval in years and the number of people vaccinated can be seen dynamically, by month.

In [5]:
import pandas as pd
import plotly.express as px
import pycountry_convert as pc

# Function to get the continent from the country code
def get_continent(country_code):
    try:
        country_alpha2 = pc.country_alpha3_to_country_alpha2(country_code)
        continent_code = pc.country_alpha2_to_continent_code(country_alpha2)
        continents = {
            'AF': 'Africa',
            'AS': 'Asia',
            'EU': 'Europe',
            'NA': 'North America',
            'OC': 'Oceania',
            'SA': 'South America'
        }
        return continents[continent_code]
    except Exception as e:
        return None

# Convert the 'date' column to date type
df['date'] = pd.to_datetime(df['date'])

# Add a continent column
df['continent'] = df['iso_code'].apply(get_continent)

# Filter the data to remove rows without a continent
df = df.dropna(subset=['continent'])

# Add a column for the month and the year
df['month_year'] = df['date'].dt.to_period('M')

# Convert 'month_year' to a string so that plotly can handle it
df['month_year_str'] = df['month_year'].astype(str)

# Group the data by continent and month
monthly_vaccinations = df.groupby(['continent', 'month_year_str'])['daily_vaccinations'].sum().reset_index()

# Create the dynamic chart
fig = px.line(monthly_vaccinations, x='month_year_str', y='daily_vaccinations', color='continent',
              labels={'daily_vaccinations': 'Dosis diarias administradas', 'month_year_str': 'Fecha', 'continent': 'Continente'},
              title='Tasa mensual de vacunación en diferentes continentes')

# Adjust the height of the chart
fig.update_layout(height=1000)

# Save the chart to an HTML file
fig.write_html(PATH+"Tasa_mensual_de_vacunación_en_diferentes_continentes.html")

# Show the chart
fig.show()



<ipython-input-5-767b91d128b3>:32: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Question 3:

Which organizations have reported the most cases of vaccinated and partially vaccinated people, and where? Stacked bar chart showing the vaccinated and partially vaccinated people, with each bar representing the total by continent and organization.

In [6]:
# Convert the 'date' column to date type
df['date'] = pd.to_datetime(df['date'])

# Filter the data to remove rows without a continent
df = df.dropna(subset=['continent'])

# Group the data by continent and organization, summing the vaccinated and partially vaccinated people
grouped_df = df.groupby(['continent', 'source_name']).agg({
    'people_vaccinated': 'sum',
    'people_fully_vaccinated': 'sum'
}).reset_index()

# Create the stacked bar chart for partially vaccinated people
fig_partially_vaccinated = px.bar(grouped_df, x='continent', y='people_vaccinated', color='source_name',
                                  labels={'people_vaccinated': 'Número de personas parcialmente vacunadas',
                                          'continent': 'Continente', 'source_name': 'Organización'},
                                  title='Total de personas parcialmente vacunadas por continente y organización',
                                  barmode='stack')

# Adjust the height of the chart
fig_partially_vaccinated.update_layout(height=2000)

# Increase the size of the X-axis labels
fig_partially_vaccinated.update_xaxes(tickfont=dict(size=20))

# Create the stacked bar chart for fully vaccinated people
fig_fully_vaccinated = px.bar(grouped_df, x='continent', y='people_fully_vaccinated', color='source_name',
                              labels={'people_fully_vaccinated': 'Número de personas completamente vacunadas',
                                      'continent': 'Continente', 'source_name': 'Organización'},
                              title='Total de personas completamente vacunadas por continente y organización',
                              barmode='stack')

# Adjust the height of the chart
fig_fully_vaccinated.update_layout(height=2000)
# Increase the size of the X-axis labels
fig_fully_vaccinated.update_xaxes(tickfont=dict(size=20))
# Show the charts
fig_partially_vaccinated.write_html(PATH+"Total_de_personas_parcialmente_vacunadas_por_continente_y_organización.html")
fig_fully_vaccinated.write_html(PATH+"Total_de_personas_completamente_vacunadas_por_continente_y_organización.html")
fig_partially_vaccinated.show()
fig_fully_vaccinated.show()

Question 4:

In which countries have the most doses been administered in relation to the country's number of inhabitants? 2 dynamic pie charts where this can be seen, for the 10 countries with the highest proportion and the 10 countries with the lowest proportion, in different colors to represent the ranking

In [7]:
# Select the 50 countries with the highest and lowest proportion of doses administered
top_countries = df.groupby('country')['total_vaccinations_per_hundred'].max().nlargest(50).reset_index()
bottom_countries = df.groupby('country')['total_vaccinations_per_hundred'].max().nsmallest(50).reset_index()

# Create the pie chart for the 50 countries with the highest proportion of doses administered
fig_top = px.pie(top_countries, names='country', values='total_vaccinations_per_hundred',
                 title='Top 50 países con mayor proporción de dosis administradas')

# Improve the readability of the chart
fig_top.update_traces(textposition='inside', textinfo='percent+label')
fig_top.update_layout(height=800, margin=dict(t=50, b=0, l=0, r=0))

# Create the pie chart for the 50 countries with the lowest proportion of doses administered
fig_bottom = px.pie(bottom_countries, names='country', values='total_vaccinations_per_hundred',
                    title='Top 50 países con menor proporción de dosis administradas')

# Improve the readability of the chart
fig_bottom.update_traces(textposition='inside', textinfo='percent+label')
fig_bottom.update_layout(height=800, margin=dict(t=50, b=0, l=0, r=0))

# Show the charts
fig_top.write_html(PATH+"Top_50_países_con_mayor_proporción_de_dosis_administradas.html")
fig_bottom.write_html(PATH+"Top_50_países_con_menor_proporción_de_dosis_administradas.html")
fig_top.show()
fig_bottom.show()

Question 5:

How consistently are the different countries reporting the data?
Graphical representation: Scatter plots or box plots showing the amount of missing data by country and date.


In [ ]:
import pandas as pd
import plotly.express as px

# Convert the 'date' column to date type
df['date'] = pd.to_datetime(df['date'])

# Calculate the total number of missing values by country and column
missing_datas = df.groupby('country').apply(lambda x: x.isnull().sum()).reset_index(drop=True)

# Add the 'country' column that was removed with reset_index(drop=True)
missing_datas['country'] = df['country'].unique()

# Filter the relevant columns and reorder to have 'country' at the beginning
columns_of_interest = ['country', 'total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated',
                       'daily_vaccinations', 'total_vaccinations_per_hundred', 'people_vaccinated_per_hundred',
                       'people_fully_vaccinated_per_hundred', 'daily_vaccinations_per_million']

missing_datas = missing_datas[columns_of_interest]

# Transform the DataFrame so that it is suitable for plotly.express
missing_data_melted = missing_datas.melt(id_vars='country',
                                         var_name='column',
                                         value_name='missing_count')

# Concatenate 'country' and 'column' to show the specific column within each country's boxplot
missing_data_melted['country_column'] = missing_data_melted['country'] + ' - ' + missing_data_melted['column']

# Create the interactive box plot
fig_box = px.box(missing_data_melted, x='country', y='missing_count',
                 points="all",
                 hover_data=['column'],
                 title='Distribución de datos faltantes por país y columna',
                 labels={'missing_count': 'Cantidad de datos faltantes', 'country': 'País', 'country_column': 'País - Columna'})

# Increase the size of the chart and add a horizontal scrollbar
fig_box.update_layout(
    height=1500,  # Increases the height of the chart
    xaxis=dict(
        tickmode='linear',
        tickangle=45,  # Tilt angle of the labels
        rangeslider=dict(visible=True)
    )
)

# Show the chart
fig_box.write_html(PATH+"Distribución_de_datos_faltantes_por_país_y_columna.html")
fig_box.show()



In [ ]:
# Convert the 'date' column to date type
df['date'] = pd.to_datetime(df['date'])

# Check the columns and the first rows of the DataFrame
print(df.columns)
print(df.head())

# Create a DataFrame to count the missing values by country and date
missing_data_counts = df.set_index(['country', 'date']).isnull().sum(axis=1).reset_index(name='missing_count')

# Filter the rows where 'missing_count' is greater than 0
missing_data_counts = missing_data_counts[missing_data_counts['missing_count'] > 0]

# Group by country and sum the missing values
total_missing_by_country = missing_data_counts.groupby('country')['missing_count'].sum().reset_index()

# Select the 30 countries with the most missing values
top_30_countries = total_missing_by_country.nlargest(30, 'missing_count')['country'].tolist()

# Filter the missing-values DataFrame to include only the 30 selected countries
top_missing_data_counts = missing_data_counts[missing_data_counts['country'].isin(top_30_countries)]

# Check the missing-count data
print(top_missing_data_counts.head())

# Create the scatter plot
fig_scatter = px.scatter(top_missing_data_counts, x='date', y='country', size='missing_count',
                         title='Cantidad de datos faltantes por país y fecha (Top 30 países)',
                         labels={'missing_count': 'Cantidad de datos faltantes', 'date': 'Fecha', 'country': 'País'},
                         hover_data=['missing_count'])

# Make sure that all 30 countries appear on the Y axis
fig_scatter.update_layout(
    yaxis=dict(
        categoryorder='array',
        categoryarray=top_30_countries
    ),
    height=800,  # Increases the height of the chart
    xaxis=dict(
        rangeselector=dict(
            buttons=list([
                dict(count=1, label="1m", step="month", stepmode="backward"),
                dict(count=6, label="6m", step="month", stepmode="backward"),
                dict(count=1, label="YTD", step="year", stepmode="todate"),
                dict(count=1, label="1y", step="year", stepmode="backward"),
                dict(step="all")
            ])
        ),
        rangeslider=dict(visible=True),
        type="date"
    ),
)

# Show the chart
fig_scatter.write_html(PATH+"Cantidad_de_datos_faltantes_por_país_y_fecha_Top_30_países.html")
fig_scatter.show()

Index(['country', 'iso_code', 'date', 'total_vaccinations',
       'people_vaccinated', 'people_fully_vaccinated', 'daily_vaccinations',
       'total_vaccinations_per_hundred', 'people_vaccinated_per_hundred',
       'people_fully_vaccinated_per_hundred', 'daily_vaccinations_per_million',
       'vaccines', 'source_name', 'year', 'continent', 'month_year',
       'month_year_str'],
      dtype='object')
       country iso_code       date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG 2021-02-22                 0.0                0.0   
1  Afghanistan      AFG 2021-02-23                 NaN                NaN   
2  Afghanistan      AFG 2021-02-24                 NaN                NaN   
3  Afghanistan      AFG 2021-02-25                 NaN                NaN   
4  Afghanistan      AFG 2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations  \
0                      NaN                 NaN   
1                      NaN   